### PR TITLE
Add support for dependent disks during in-cluster near-live migration

### LINKED ISSUE #3059: Near-live migration for containers
### Is there an existing issue for this?

- [x] There is no existing issue for this feature

### What are you currently unable to do

Currently, there isn't a good solution for moving large containers quickly and with low downtime (other than getting better network cards or a better internet plan lol)

### What do you think would need to be added

Something along the lines of:

snapshot  the container
send initial snapshot to target (e.g. btrfs send/receive)
repeatedly take incremental snapshots and keep sending deltas until the remaining diff is small
stop the container
take final snapshot and send it over
start the container up on the target host

This should, based on my (admittedly janky) testing, significantly reduce downtime compared to a full shutdown-move-start loop, and also would work on instances with things that CRIU doesn't support, like instances with network interfaces)

I’ve only tested this kind of workflow manually with btrfs, but I assume similar approaches could work with other backends that support incremental send (e.g. ZFS).

### PR DESCRIPTION
**Pull request by presztak**
Closes: #3059

### CODE REVIEW

**Comment by stgraber** [comment]
@presztak I really don't like the added shared/api options. We shouldn't have anything internal be exposed to the public API in that way.

Is there no way for us to detect what we need internally considering all of this is happening with a shared database and with the client user-agent and certificate set to the cluster internal one so we can detect that this isn't a user request but is in fact a cluster-internal operation.

**Comment by presztak** [comment]
@stgraber Another option I thought, would be to detect that this is a `copy` request originating from a different node, I mean, an inter-cluster request. However, reusing `copy` request elsewhere would then cause us to always skip shared volumes.

**Comment by presztak** [comment]
quick update, on the server side, we're now using `migration` directly instead of `copy`. This avoids exposing new external fields related to dependent disk migration, and has reduced the number of commits to 4. I've moved the PR back to draft, since we're still missing support for near-live migration when the remote is on shared storage but dependent disks are local.

File: cmd/incusd/instance_post_nearlive.go [hunk 1]
```diff
@@ -29,20 +33,23 @@ func nearLiveMigrationSupported(s *state.State, inst instance.Instance, req api.
...
-	dependent := false
 	err := inst.ForEachDependentDiskType(func(dev deviceConfig.DeviceNamed) error {
-		dependent = true
+		diskPool, err := storagePools.LoadByName(s, dev.Config["pool"])
+		if err != nil {
+			return fmt.Errorf("Failed loading storage pool: %w", err)
+		}
+
+		driverName := diskPool.Driver().Info().Name
+		if driverName == "dir" || driverName == "lvm" {
```
Comment (stgraber) [inline cmd/incusd/instance_post_nearlive.go]: Can we please make that a DriverInfo property so we don't need to hardcode drivers name.

File: test/suites/clustering_move.sh [hunk 2]
```diff
@@ -195,8 +195,31 @@ EOF
         INCUS_DIR="${INCUS_ONE_DIR}" incus info c4 | grep -q "Location: node2"
         INCUS_DIR="${INCUS_ONE_DIR}" incus info c4 | grep -q "Status: STOPPED"
 
+        # Dependent volumes are moved along with the instance.
```
Comment (stgraber) [inline test/suites/clustering_move.sh:289]: Can we add a test running on ceph and linstor which validates the tricky edge cases?

Specifically, I'd like to see two more tests:
 - Instance on shared storage, one dependent volume on local, one dependent volume on shared, one regular volume on shared
 - Instance on local storage, one dependent volume on local, one dependent volume on shared, one regular volume on shared

I'd also prefer that the test writes a basic test file to the instance and to each of the dependent and regular volumes just prior to the move and confirms that the files exist with the same content following the move.

File: cmd/incusd/instance_post_nearlive.go (hunk 1 above)
Comment (presztak) [reply cmd/incusd/instance_post_nearlive.go]: Make sense. Done.

**Comment by stgraber** [comment]
@presztak what's with that "test" commit? :)

**Comment by presztak** [comment]
@stgraber it's just for testing purpose for now as I've hit issues on linstor driver.

File: test/suites/clustering_move.sh (hunk 2 above)
Comment (presztak) [reply test/suites/clustering_move.sh:289]: Tests added. One test for linstor is failing so we need to wait for: https://github.com/lxc/incus/pull/3866

File: internal/server/storage/drivers/driver_cephfs.go [hunk 3]
```diff
@@ -94,6 +94,7 @@ func (d *cephfs) Info() Info {
 		OptimizedImages:              false,
 		PreservesInodes:              false,
 		Remote:                       d.isRemote(),
+		NearLiveMigration:            true,
```
Comment (stgraber) [inline internal/server/storage/drivers/driver_cephfs.go]: instances can't be on cephfs, so that should be a no

File: internal/server/storage/drivers/driver_truenas.go [hunk 4]
```diff
@@ -105,6 +105,7 @@ func (d *truenas) Info() Info {
 		OptimizedBackups:             false,
 		PreservesInodes:              false,
 		Remote:                       d.isRemote(),
+		NearLiveMigration:            true,
```
Comment (stgraber) [inline internal/server/storage/drivers/driver_truenas.go]: Should be false as we don't have a good server to server incremental mechanism.

File: internal/server/storage/drivers/driver_lvm.go [hunk 5]
```diff
@@ -156,6 +156,7 @@ func (d *lvm) Info() Info {
 		OptimizedImages:              d.usesThinpool(), // Only thinpool pools support optimized images.
 		PreservesInodes:              false,
 		Remote:                       d.isRemote(),
+		NearLiveMigration:            d.clustered,
```
Comment (stgraber) [inline internal/server/storage/drivers/driver_lvm.go]: Should be false as we don't have a good server to server incremental mechanism.

File: internal/server/storage/drivers/driver_linstor.go [hunk 6]
```diff
@@ -294,6 +294,7 @@ func (d *linstor) Info() Info {
 		DefaultVMBlockFilesystemSize: deviceConfig.DefaultVMBlockFilesystemSize,
 		Buckets:                      false,
 		Remote:                       d.isRemote(),
+		NearLiveMigration:            true,
```
Comment (stgraber) [inline internal/server/storage/drivers/driver_linstor.go]: Should be false as we don't have a good server to server incremental mechanism.

File: internal/server/storage/drivers/driver_ceph.go [hunk 7]
```diff
@@ -93,6 +93,7 @@ func (d *ceph) Info() Info {
 		OptimizedImages:              true,
 		PreservesInodes:              false,
 		Remote:                       d.isRemote(),
+		NearLiveMigration:            true,
```
Comment (stgraber) [inline internal/server/storage/drivers/driver_ceph.go]: Should be false as we don't have a good server to server incremental mechanism.

**Comment by stgraber** [comment]
@presztak the Info flag seems wrong to me. It's meant to indicate whether an efficient transfer of incremental state is available or not (the requirement for near live migration).

Only ZFS and btrfs offer that. The rest all rely on rsync/dd which aren't suitable for this use case.

Remote pools within a cluster also don't need the flag set to true anyway since the data isn't going to be moving at all. But if you were to go cross-cluster between remote storage drivers on both ends, none of our drivers have an efficient way to transfer those snapshots so they shouldn't be treated as handling near live migration.

**Comment by presztak** [comment]
@stgraber, ok, I did it that way because the initial condition was:
```
if poolInfo.Name == "dir" || poolInfo.Name == "lvm" {
    return fmt.Errorf("Near-live migration isn't supported for dependent disk %q on %q storage pools", dev.Name, poolInfo.Name)
}
```
So `NearLiveMigration` was `true` for everything except `dir` and `lvm`. But I see your point now, and I've made adjustments based on your feedback.

File: internal/server/storage/drivers/driver_cephfs.go (hunk 3 above)
Comment (presztak) [reply internal/server/storage/drivers/driver_cephfs.go]: Fixed.

File: internal/server/storage/drivers/driver_truenas.go (hunk 4 above)
Comment (presztak) [reply internal/server/storage/drivers/driver_truenas.go]: Fixed.

File: internal/server/storage/drivers/driver_lvm.go (hunk 5 above)
Comment (presztak) [reply internal/server/storage/drivers/driver_lvm.go]: Fixed.

File: internal/server/storage/drivers/driver_linstor.go (hunk 6 above)
Comment (presztak) [reply internal/server/storage/drivers/driver_linstor.go]: Fixed.

File: internal/server/storage/drivers/driver_ceph.go (hunk 7 above)
Comment (presztak) [reply internal/server/storage/drivers/driver_ceph.go]: Fixed.
